File: src/libs/cpp/actions/push_exception.cpp
#include "push_exception.hpp"

void push_exception(VirtualMachine &vm) noexcept
{
	auto &instruction = vm.instructions[vm.instruction_index];
	Value arg;

	if (instruction.operand[1])
	{
		arg = vm.stack.pop();
	}
	else
	{
		arg = vm.get_const(instruction.operand[0]);
	}

	std::map<std::string, Value> err;
	err["message"] = vm.stack.pop();
	err["stack"] = std::vector<Value>{instruction.line_no};
	err["type"] = arg;
	err["file"] = "unknown";
	err["line"] = instruction.line_no;

	vm.stack.push(err);
}

File: src/runtime/actions/push_exception.lua
return function(vm, line, p1, p2)
	--Build the exception object to handle or throw later.
	local err = std.object()
	err.message = vm.pop()
	err.stack = { line }
	err.type = p1
	err.file = FILE
	err.line = line

	vm.push(err)
end
